### PR TITLE
Fix validation in TdmProgram.compile

### PR DESCRIPTION
**Pull request by lneuhaus**
**Context:**
We recently changed the allowed gate parameters for `TD2`. The `Rgate` phase can now be both positive and negative. With this change, requested programs did not pass the validation step in the program validation any more when negative phases were set. The reason was that when the code was supposed to validate the phases of the `BSgate` (where only positive values are allowed), it used the values set for the `Rgate` phases, which were occasionally negative. 

**Description of the Change:**
The proposed solution now uses the correct values in the above described validation step for each parameter.

**Benefits:**
Fixed a bug. Code behavior should be closer to what one expects now.

**Possible Drawbacks:**
I did not fully understand the logic. In particular, it seems to me that the current way of updating a counter for the index that is used to grab the relevant parameter values is highly error-prone. I could imagine there are edge cases where this still leads to unexpected behavior. A more thorough logic cleanup would maybe be a better solution, but take more time. I also did not dig too far into why the tests did not uncover this problem. I'd be happy to add a new test case in case it's worth it and we don't want to do the full cleanup now.

**Related GitHub Issues:**
None

### CODE REVIEW

File: strawberryfields/tdm/tdmprogram.py [hunk 1]
```diff
@@ -471,6 +471,8 @@ def compile(self, *, device=None, compiler=None):
             # Third check: the parameters of the gates are valid
             # We will loop over the different operations in the device specification
 
+            num_symbolic_param = 0  # counts the number of symbolic variables, which are labelled consecutively by the context method
```
Comment (lneuhaus) [inline strawberryfields/tdm/tdmprogram.py:474]: the issue was line 512 (`local_p_vals = self.tdm_params[num_symbolic_param]`): before this change, the index was effectively reset to zero for every gate, thus picking the wrong parameter.

**Review by fab1an-q — APPROVED** [review]

**Review by thisac — COMMENTED** [review]
Nice catch @lneuhaus. 

This could potentially be slightly improved by using the `TDMProgram.parameters` method instead, which should return the same as the values in `TDMProgram.tdm_params`, which is a dictionary with `p0`, `p1`, etc. as keys. The line 

```
local_p_vals = self.tdm_params[num_symbolic_param]
``` 

could in that case be changed to 

```
local_p_vals = self.parameters.get(program_param.name, [])
```

and `num_symbolic_param` could be removed altogether.

File: strawberryfields/tdm/tdmprogram.py [hunk 2]
```diff
@@ -580,7 +580,7 @@ def unroll(self, shots):
         q = self.register
 
         sm = []
-        for i in range(len(self.N)):
+        for i, _ in enumerate(self.N):
```
[the remark this answers is not in the document]
Comment (lneuhaus) [reply strawberryfields/tdm/tdmprogram.py:583]: I would not have touched this line vy default, but codeFactor (one of the CI tests) was failing because of the old version of this line. Happy to revert to the original if you help me get past the failed codeFactor check.

[the remark this answers is not in the document]
Comment (thisac) [reply strawberryfields/tdm/tdmprogram.py:583]: I'm actually not sure which I prefer. I would have used the former (which I believe might be slightly faster, but not by much), although I actually now thinking that using `enumerate` looks cleaner. I'm happy to keep your change here.

**Review by thisac — APPROVED** [review]
Updated with the suggested fix above. It might be good adding a test for this specific case as well now, just to make sure that everything works as it should.

**Comment by lneuhaus** [comment]
@thisac I added a few tests to make sure the validation works as one would expect. I checked that these tests fail on the current master branch, so am confident they cover the bug that we're fixing here.

**Comment by lneuhaus** [comment]
The failed codecov-check seems to be unrelated to this PR - I will assume someone will fix this in another branch and probably just wait for that solution?

File: tests/frontend/test_tdmprogram.py [hunk 3]
```diff
@@ -756,3 +758,68 @@ def test_shots_passed(self):
...
+        self.compile_test_program(device, args=(-1, 1, 2, 3))
+
+    @pytest.mark.parametrize("incorrect_index", list(range(4)))
+    def test_validation_incorrect_args(self, device, incorrect_index):
+        """Test the correct error is raised when the tdm circuit explicit parameters are not within the allowed ranges"""
+        args = [-1, 1, 2, 3]
+        args[incorrect_index] = -999
+        with pytest.raises(CircuitError, match="Parameter has value '-999' while its valid range is "):
+            self.compile_test_program(device, args=args)
+
```
Comment (thisac) [inline tests/frontend/test_tdmprogram.py:825]: There seems to be a slight issue here with the ending newline. Perhaps just remove the final tab and make sure there's only a single empty line at the end? 🤔

**Review by thisac — APPROVED** [review]
Looks great @lneuhaus! 💯 

Just the final newline that looks a bit weird (GitHub seems to be warning about it), but otherwise looks fine.
